### ISSUE TITLE
Problem using CASE WHEN on ordering or filtering 

### ISSUE THREAD

**Issue by ptykamikaze** [yajra/laravel-datatables#115]
I have the a query that returns a (case when) column.. How can I override the column name with the (case when) definition because it is throwing me errors when sorting or filtering by the (case when) column

**Comment by yajra** [comment]
You can override the default global search for specific column by using `filterColumn` function. However, for custom sorting, we currently don't have function to override ordering. See this example for filterColumn. http://datatables.yajrabox.com/eloquent/post-column-search

On the other hand, `Datatables` class extends query builder functionality. This will let you do your own order method like below:

```php
....
use yajra\Datatables\Datatables;
....

public function getData(Datatables $datatables)
{
    $users = User::select(['id', 'name', 'email', 'created_at', 'updated_at']);

    $datatables = $datatables->eloquent($users);
    if ($datatables->request->isColumnOrderable(1)) {
        $datatables->orderByRaw('case when ...');
    }

    return $datatables->make(true);
}

```

> Note: For column 1 js definition, you should use a valid name like `{ data: 'case_when_data', 'name:'id' }`. The proposed solution above is just a work around.

Lastly, I think I will add a new feature to override the default global ordering like `orderColumn`. Thanks!

**Comment by ptykamikaze** [comment]
I was trying to solve this by myself so I added an array variable to store the column name and the (case when) definition. for Example: alias_cols['column_1'] = 'CASE WHEN .... END'.

So I added this to the columnSearch and compileGlobalSearch functions

if( isset($this->aliases_columns[$column])) {
 $this->query->whereRaw('LOWER(' . $this->alias_cols[$column] . ') LIKE ?', [Str::lower($keyword)]);
}

It is working fine when filtering... however, while ordering by this column, the query returns no DATA and if I take the query and run it in mysql workbench, it returns data...

The ordering function was changed like this...

        foreach ($this->request->orderableColumns() as $orderable) {
            $column = $this->getColumnName($orderable['column']);
			if( isset($this->aliases_columns[$column])) {
				$this->query->orderByRaw($this->aliases_columns[$column], $orderable['direction']);
			}
			else
            	$this->query->orderBy($column, $orderable['direction']);
        }

any ideas why it is returning no data?

**Comment by yajra** [comment]
@ptykamikaze, I think what you need to do is just use the alias on your js name definition like `{ data: 'alias', name: 'alias' }`. On mysql, ordering using alias is allowed. Not sure though on other DB.

For filtering, you will still need to use `filterColumn` since the `alias` column does not exists.

**Comment by ptykamikaze** [comment]
Thanks for your help!

**Comment by ptykamikaze** [comment]
The filterColumn works fine with individual column search but with the main search is not working.

When using main search, the query is built like this SELECT * FROM table WHERE (CASE WHEN) LIKE ? AND .....other conditions....

When using individual column, the query is built like this SELECT * FROM table WHERE (CASE WHEN) LIKE ?

I see the main search is not including the filterColumn within the OR conditions

**Comment by yajra** [comment]
`filterColumn` should be included inside the OR conditions because it's part of the global search like `select * from table where (a like ? OR b like ?)`. If not, then there's a bug. But as far as I know it works well. Can't verify though atm.

**Comment by ptykamikaze** [comment]
The link you sent me has the bug as well ... http://datatables.yajrabox.com/eloquent/post-column-search if you put in the main search "10", the table returns nothing, even if there are records with 10, 100-100, 101-101, etc

**Comment by yajra** [comment]
@ptykamikaze, have confirmed the bug. Will post a patch later today to fix it. Thanks!

**Comment by ptykamikaze** [comment]
Thank you for the fast response!

**Comment by yajra** [comment]
@ptykamikaze, filterColumn was re-implemented and much simple to use now. I have added a special variable `$1` which will correspond to actual search value. See updated demo http://datatables.yajrabox.com/eloquent/post-column-search for usage. Should be fixed on v5.6.0
